File: packages/ui-swiftui/Sources/KinetixUI/ComparisonSlider.swift
//
// ComparisonSlider.swift — KinetixComparisonSlider.
//
// Mirrors packages/ui/src/components/comparison-slider.tsx: a drag
// handle wiping between two stacked layers. Unlike KinetixSlider (which
// wraps the system Slider), this needs a fully custom visual, so it's
// built on a plain DragGesture over a GeometryReader-measured width —
// the "after" layer is masked with a transparent-then-opaque HStack
// (rather than the alignment-parameterized `.mask`, to avoid an
// availability question) instead of Compose's `clipRect` or the web's
// CSS `clip-path`. Only dragging the handle is supported — the web
// version's Radix Slider also lets you click anywhere on the track to
// jump, a documented scope-down, not a silent gap.
//

import SwiftUI

public struct KinetixComparisonSlider<Before: View, After: View>: View {
    @Environment(\.kinetixColors) private var colors

    private let before: Before
    private let after: After
    private let beforeLabel: String?
    private let afterLabel: String?
    private let externalValue: Binding<Double>?

    @State private var internalValue: Double

    public init(
        value: Binding<Double>? = nil,
        defaultValue: Double = 50,
        beforeLabel: String? = nil,
        afterLabel: String? = nil,
        @ViewBuilder before: () -> Before,
        @ViewBuilder after: () -> After
    ) {
        self.externalValue = value
        self._internalValue = State(initialValue: value?.wrappedValue ?? defaultValue)
        self.beforeLabel = beforeLabel
        self.afterLabel = afterLabel
        self.before = before()
        self.after = after()
    }

    private var current: Double {
        externalValue?.wrappedValue ?? internalValue
    }

    private func setCurrent(_ v: Double) {
        let clamped = min(100, max(0, v))
        if let externalValue {
            externalValue.wrappedValue = clamped
        } else {
            internalValue = clamped
        }
    }

    @ViewBuilder
    private func labelChip(_ text: String) -> some View {
        Text(text)
            .font(.kinetixLabelSm)
            .foregroundStyle(colors.foreground)
            .padding(.horizontal, 8)
            .padding(.vertical, 2)
            .background(colors.background.opacity(0.8), in: RoundedRectangle(cornerRadius: 4, style: .continuous))
            .padding(8)
    }

    public var body: some View {
        GeometryReader { proxy in
            let width = proxy.size.width
            let handleX = width * CGFloat(current / 100)

            ZStack(alignment: .topLeading) {
                before
                after
                    .mask(
                        HStack(spacing: 0) {
                            Color.clear.frame(width: handleX)
                            Color.black
                        }
                    )

                Rectangle()
                    .fill(colors.background)
                    .frame(width: 2)
                    .offset(x: handleX - 1)

                if let beforeLabel {
                    labelChip(beforeLabel)
                }
                if let afterLabel {
                    labelChip(afterLabel)
                        .frame(maxWidth: .infinity, alignment: .trailing)
                }

                Circle()
                    .fill(colors.background.opacity(0.9))
                    .overlay(Circle().strokeBorder(colors.background, lineWidth: 2))
                    .frame(width: 32, height: 32)
                    .offset(x: handleX - 16, y: proxy.size.height / 2 - 16)
                    .contentShape(Circle())
                    .gesture(
                        DragGesture(minimumDistance: 0)
                            .onChanged { g in
                                setCurrent(Double(g.location.x / width) * 100)
                            }
                    )
            }
            .clipShape(RoundedRectangle(cornerRadius: 8, style: .continuous))
        }
        .aspectRatio(16.0 / 9.0, contentMode: .fit)
    }
}

File: packages/ui-flutter/lib/src/comparison_slider.dart
import 'package:flutter/material.dart';

import 'app_text.dart';
import 'theme.dart';

/// Mirrors `packages/ui/src/components/comparison-slider.tsx`: a drag
/// handle wiping between two stacked layers. Unlike [KinetixSlider]
/// (which wraps the system `Slider`), this needs a fully custom visual,
/// so it's built on a plain `GestureDetector` + a `CustomClipper<Rect>`
/// revealing the `after` layer from the handle position onward — the
/// same "duplicate the layers, clip one of them" technique as the other
/// two native ports and the web's CSS `clip-path`. Both dragging the
/// handle and tapping anywhere on the track jump the position (Flutter's
/// gesture layer makes both cheap here, unlike the Compose/SwiftUI ports
/// which only drag the handle).
class KinetixComparisonSlider extends StatefulWidget {
  const KinetixComparisonSlider({
    super.key,
    required this.before,
    required this.after,
    this.value,
    this.defaultValue = 50,
    this.onValueChanged,
    this.beforeLabel,
    this.afterLabel,
  });

  final Widget before;
  final Widget after;
  final double? value;
  final double defaultValue;
  final ValueChanged<double>? onValueChanged;
  final String? beforeLabel;
  final String? afterLabel;

  @override
  State<KinetixComparisonSlider> createState() => _KinetixComparisonSliderState();
}

class _KinetixComparisonSliderState extends State<KinetixComparisonSlider> {
  late double _internal = widget.value ?? widget.defaultValue;

  double get _current => widget.value ?? _internal;

  void _setCurrent(double v) {
    final clamped = v.clamp(0.0, 100.0);
    setState(() => _internal = clamped);
    widget.onValueChanged?.call(clamped);
  }

  Widget _labelChip(BuildContext context, String text) {
    final c = KinetixTheme.of(context);
    return Container(
      margin: const EdgeInsets.all(8),
      padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
      decoration: BoxDecoration(
        color: c.background.withValues(alpha: 0.8),
        borderRadius: BorderRadius.circular(4),
      ),
      child: Text(text, style: AppText.labelSm.copyWith(color: c.foreground)),
    );
  }

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);

    return AspectRatio(
      aspectRatio: 16 / 9,
      child: ClipRRect(
        borderRadius: BorderRadius.circular(8),
        child: LayoutBuilder(
          builder: (context, constraints) {
            final width = constraints.maxWidth;
            final handleX = width * _current / 100;

            return GestureDetector(
              onHorizontalDragUpdate: (details) => _setCurrent((details.localPosition.dx / width) * 100),
              onTapDown: (details) => _setCurrent((details.localPosition.dx / width) * 100),
              child: Container(
                color: c.muted,
                child: Stack(
                  children: [
                    Positioned.fill(child: widget.before),
                    ClipRect(
                      clipper: _RevealClipper(handleX),
                      child: SizedBox.expand(child: widget.after),
                    ),
                    Positioned(
                      left: handleX - 1,
                      top: 0,
                      bottom: 0,
                      child: Container(width: 2, color: c.background),
                    ),
                    if (widget.beforeLabel != null)
                      Positioned(top: 0, left: 0, child: _labelChip(context, widget.beforeLabel!)),
                    if (widget.afterLabel != null)
                      Positioned(top: 0, right: 0, child: _labelChip(context, widget.afterLabel!)),
                    Positioned(
                      left: handleX - 16,
                      top: 0,
                      bottom: 0,
                      child: Center(
                        child: Container(
                          width: 32,
                          height: 32,
                          decoration: BoxDecoration(
                            shape: BoxShape.circle,
                            color: c.background.withValues(alpha: 0.9),
                            border: Border.all(color: c.background, width: 2),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}

class _RevealClipper extends CustomClipper<Rect> {
  const _RevealClipper(this.left);

  final double left;

  @override
  Rect getClip(Size size) => Rect.fromLTWH(left, 0, size.width - left, size.height);

  @override
  bool shouldReclip(covariant _RevealClipper oldClipper) => oldClipper.left != left;
}
